pubsublite/internal/wire/subscriber.go

```diff
@@ -0,0 +1,321 @@
+	return s.subClient.Subscribe(addSubscriptionRoutingMetadata(ctx, s.subscription))
+}
+
+func (s *subscribeStream) initialRequest() (interface{}, bool) {
+	return s.initialReq, true
+}
+
+func (s *subscribeStream) validateInitialResponse(response interface{}) error {
+	subscribeResponse, _ := response.(*pb.SubscribeResponse)
+	if subscribeResponse.GetInitial() == nil {
+		return errInvalidInitialSubscribeResponse
+	}
+	return nil
+}
+
+func (s *subscribeStream) onStreamStatusChange(status streamStatus) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	switch status {
+	case streamConnected:
+		s.unsafeUpdateStatus(serviceActive, nil)
+
+		// Reinitialize the offset and flow control tokens when a new subscribe
+		// stream instance is connected.
+		if seekReq := s.offsetTracker.RequestForRestart(); seekReq != nil {
+			if s.stream.Send(&pb.SubscribeRequest{
+				Request: &pb.SubscribeRequest_Seek{Seek: seekReq},
+			}) {
+				s.seekInFlight = true
+			}
+		}
+		s.unsafeSendFlowControl(s.flowControl.RequestForRestart())
+		s.pollFlowControl.Start()
+
+	case streamReconnecting:
+		s.seekInFlight = false
+		s.pollFlowControl.Stop()
+
+	case streamTerminated:
+		s.unsafeInitiateShutdown(serviceTerminated, s.stream.Error())
+	}
+}
+
+func (s *subscribeStream) onResponse(response interface{}) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.status >= serviceTerminating {
+		return
+	}
+
+	processResponse := func() error {
+		subscribeResponse, _ := response.(*pb.SubscribeResponse)
+		switch {
+		case subscribeResponse.GetMessages() != nil:
+			return s.unsafeOnMessageResponse(subscribeResponse.GetMessages())
+		case subscribeResponse.GetSeek() != nil:
+			return s.unsafeOnSeekResponse(subscribeResponse.GetSeek())
+		default:
+			return errInvalidSubscribeResponse
+		}
+	}
+	if err := processResponse(); err != nil {
+		s.unsafeInitiateShutdown(serviceTerminated, err)
+	}
+}
+
+func (s *subscribeStream) unsafeOnSeekResponse(response *pb.SeekResponse) error {
+	if !s.seekInFlight {
+		return errNoInFlightSeek
+	}
+	s.seekInFlight = false
+	return nil
+}
+
+func (s *subscribeStream) unsafeOnMessageResponse(response *pb.MessageResponse) error {
+	if len(response.Messages) == 0 {
+		return errServerNoMessages
+	}
+	if err := s.offsetTracker.OnMessages(response.Messages); err != nil {
+		return err
+	}
+	if err := s.flowControl.OnMessages(response.Messages); err != nil {
+		return err
+	}
+	for _, msg := range response.Messages {
+		// Register outstanding acks, which are primarily handled by the
+		// `committer`.
+		ack := newAckConsumer(msg.GetCursor().GetOffset(), msg.GetSizeBytes(), s.onAck)
+		if err := s.acks.Push(ack); err != nil {
+			return err
+		}
+
+		// Release the mutex before delivering the message to the receiver. This
+		// allows the user to stop the subscriber (which would discard remaining
+		// messages) and other operations (e.g. ack processing, batch flow control)
+		// to occur.
+		s.mu.Unlock()
```

Comment:
This means that we are locking/unlocking this mutex once per message. Are we worried about the overhead? Would it be better to gather all messages and send them to the client in a batch?

Reply:
Actually, I've thought of a way to deliver a batch of messages to the shim layer (which delivers single messages to users) and avoid the per-message locking/unlocking here. If it's a relatively minimal change, I'll update this PR. Otherwise will submit as-is and ament in a subsequent PR.

Thanks for pointing this out!

Reply:
The change was minimal, so I just amended this PR.